src/Service/DefaultBlockInstaller.php

```diff
@@ -0,0 +1,185 @@
+        unset($themes[$i]);
+      }
+    }
+
+    return $themes;
+  }
+
+  /**
+   * Gets the name of the active theme, if there is one.
+   */
+  protected function getActiveThemeName(): ?string {
+    $activeTheme = $this->themeManager->getActiveTheme()->getName();
+    if ($activeTheme === 'core') {
+      // 'core' is what Drupal returns when there's no themes available.
+      // See \Drupal\Core\Theme\ThemeInitialization::getActiveThemeByName().
+      // This mainly happens in the installer.
+      return NULL;
+    }
+    return $activeTheme;
+  }
+
+  /**
+   * Installs the default blocks for the given modules.
+   *
+   * If the site is configured not to allow default blocks to be installed, this
+   * method will do nothing.
+   */
+  public function install(array $modules): void {
+
+    // If localgov_core.settings.install_default_blocks is set to FALSE, don't
+    // install default blocks. This lets site owners opt out if desired.
+    $config = $this->configFactory->get('localgov_core.settings');
+    if ($config && ($config->get('install_default_blocks') === FALSE)) {
+      return;
+    }
+
+    foreach ($modules as $module) {
+      $this->installForModule($module);
+    }
+  }
+
+  /**
+   * Installs the default blocks for the given module.
+   */
+  protected function installForModule(string $module): void {
+
+    $blocks = $this->blockDefinitions($module);
+
+    // Loop over every theme and block definition, so we set up all the blocks
+    // in all the relevant themes.
+    foreach ($this->targetThemes() as $theme) {
+      foreach ($blocks as $block) {
+
+        // Verify that the theme we're using has the requested region.
+        if (!$this->themeHasRegion($theme, $block['region'])) {
+          continue;
+        }
+
+        $block['id'] = $this->sanitiseId($theme . '_' . $block['plugin']);
+        $block['theme'] = $theme;
+
+        $this->entityTypeManager
+          ->getStorage('block')
+          ->create($block)
+          ->save();
+      }
+    }
+  }
+
+  /**
+   * Replace characters that aren't allowed in config IDs.
+   *
+   * This is partly based on
+   * \Drupal\Core\Block\BlockPluginTrait::getMachineNameSuggestion().
+   */
+  protected function sanitiseId(string $id): string {
+
+    // Shift to lower case.
+    $id = mb_strtolower($id);
+
+    // Limit to alphanumeric chars, dot and underscore.
+    $id = preg_replace('@[^a-z0-9_.]+@', '_', $id);
+
+    // Remove non-alphanumeric chars from the beginning and end of the id.
+    $id = preg_replace('@^([^a-z0-9]+)|([^a-z0-9]+)$@', '', $id);
+
+    return $id;
+  }
+
+  /**
+   * Does the given theme have the given region?
+   */
+  protected function themeHasRegion(string $theme, string $region): bool {
+    return in_array($region, $this->themeRegions($theme));
+  }
+
+  /**
+   * Gets the regions for the given theme.
+   */
+  protected function themeRegions(string $theme): array {
+    if (!isset($this->themeRegions[$theme])) {
```

Comment:
getTheme() throws an exception if the theme does not exist so check first, e.g.: 
```
if (!$this->themeHandler->themeExists($theme)) {
  return [];
}

Reply:
The list of themes that's used comes from DefaultBlockInstaller::targetThemes(), which already checks the themes exist.

Reply:
oh yeah